src/test/ui/rfc-2093-infer-outlives/issue-102966.rs

```diff
@@ -0,0 +1,44 @@
+// check-pass
+//
+// This is the example from implicit_infer.rs.
+//
+// Without cycle detection, we get:
+//
+// | Round | New predicates                                                                 |
+// | ----- | -------------------------------------------------------------------------------|
+// |   (0) | Var has explicit bound R : 'var                                                |
+// |     1 | Node gets <T as Trait<'node>>::Assoc: 'node                                    |
+// |     1 | Var  gets <RGen<R> as Trait<'var>>::Assoc: 'var                                |
+// |     2 | Node gets <RGen<<T as Trait<'node>>::Assoc> as Trait<'node>>::Assoc 'node      |
+// |     2 | Var  gets <RGen<<RGen<R> as Trait<'var>>::Assoc> as Trait<'var>>::Assoc: 'var  |
+// |   3.. | Goes on forever.                                                               |
```

Comment:
I don't think we should allow this code in the way you did it. This is probably unsound in case `<RGen<R> as Trait<'var>>::Assoc` does not resolve to `R`. I can check out your branch and play with a few examples to try to come up with a real bug case, but I'd prefer to error in case of a cycle and try to resolve cycles by normalizing opportunistically

Reply:
I'm happy to PR an error bailout if you want. Erroring is the conservative approach in that it does not make anything compile that previously didn't, before we're sure that's ok.

I think you'll find it's sound, though. As I understand it, inferred outlives bounds are a convenience, that relieves you from writing the bounds yourself. The only things the cycle detector can do wrong are:

- Kick in "too early", i.e. not adding enough bounds. Basically, if you fail to add the necessary bounds, *rustc is still checking your structs internally for whether the bounds of the types you use are met*.  The Rust reference type `&'a T` has a bound on it relating the referent type and the lifetime. You have to prove this any time you use it, and adding a `T: 'a` bound to `struct A<'a, T>(&'a T)` is just kicking this down the road to whoever uses `A`. Before RFC2093, the kicking of that can down the road was not automatic, and so writing that struct definition would result in the `&'a T` type complaining its bounds weren't met, and you having to writing some manual bounds. That's the worst case scenario.
- Kick in "too late", i.e. add too many bounds. This is definitionally sound. You can never produce unsoundness by adding more constraints on the lifetime of a type parameter. Maybe the resulting type is less versatile than it could be, but it won't be unsound.

For posterity, I did think about the whole "what if `RGen<R>::Assoc` is not `R` thing, and it is a red herring. The equivalence is not how the cycle was formed: you can change it to

```rust
trait Trait<'a> { type Assoc; /* no bound */ }
impl<'a, R> Trait<'a> for RGen<R> { type Assoc = String }
```

... and there is still a cycle and rustc hangs, because the cycle comes from the repeated substitution of `Var`'s `R: 'var` through projections that create more names for the same types. The solution to the "what outlives bounds do we need to add to satisfy the types inside `Node` and `Var`" problem is just "`Node` gets `T::Assoc: 'node`". Don't trust me, trust a pre-RFC2093 compiler (1.12.1):

```
error[E0309]: the associated type `<T as issue::Trait<'node>>::Assoc` may not live long enough
 --> src/issue.rs:5:5
  |
5 |     var: Var<'node, T::Assoc>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: consider adding an explicit lifetime bound `<T as issue::Trait<'node>>::Assoc: 'node`...
note: ...so that the type `<T as issue::Trait<'node>>::Assoc` will meet its required lifetime bounds
 --> src/issue.rs:5:5
  |
5 |     var: Var<'node, T::Assoc>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
```

When you add that bound, rustc-1.12.1 is happy. It turns out that only one additional bound was necessary to infer here, so my propagation solution exhibits the problem of "kicking in too late". Perhaps I should revisit the choice not to add the site of the original explicit bound as part of the provenance.

Reply:
Ohhhh, this is how it can be unsound: https://github.com/rust-lang/rust/issues/84305. I get you now. I still think this will turn out sound but yeah, I will PR a simple recursion limit first.